### ISSUE TITLE
Kill signals escape containers up to the parent process 

### ISSUE THREAD

**Issue by kcgthb** [apptainer/singularity#1432]
## Version of Singularity:

``` $ singularity --version
2.4.5-dist
```

### Expected behavior

Kill signals don't escape containers.

### Actual behavior

Kill signals escape containers.

### Steps to reproduce behavior

I'm connected on a remote computer (via SSH or a job scheduler):
```
$ hostname
sh-06-34.int
```
I import/run start https://hub.docker.com/r/rocker/rstudio/
```
$ singularity run docker://rocker/rstudio
Docker image path: index.docker.io/rocker/rstudio:latest
Cache folder set to /home/users/kilian/.singularity/docker
[6/6] |===================================| 100.0%
Creating container runtime...
[fix-attrs.d] applying owners & permissions fixes...
[fix-attrs.d] 00-runscripts: applying...
s6-chown: fatal: unable to chown /etc/cont-init.d/userconf: Operation not permitted
s6-chown: fatal: unable to chown /etc/cont-init.d/add: Operation not permitted
s6-chown: fatal: unable to chown /etc/fix-attrs.d/00-runscripts: Operation not permitted
s6-chown: fatal: unable to chown /etc/services.d/rstudio/run: Operation not permitted
s6-chown: fatal: unable to chown /etc/services.d/rstudio/finish: Operation not permitted
[fix-attrs.d] 00-runscripts: exited 0.
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] add: executing...
Nothing additional to add
[cont-init.d] add: exited 0.
[cont-init.d] userconf: executing...
usermod: user 'kilian' already exists
groupmod: failure while writing changes to /etc/group
usermod: failure while writing changes to /etc/group
chown: invalid group: ‘kilian:kilian’
USER is now kilian
Changing password for kilian.
chpasswd: (user kilian) pam_chauthtok() failed, error:
Authentication token manipulation error
chpasswd: (line 1, user kilian) password not changed
[cont-init.d] userconf: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
```
The containers run, all is good.

Now, I type Ctrl-C *in the container*:

```
^C[cont-finish.d] executing container finish scripts...
[cont-finish.d] done.
[s6-finish] syncing disks.
[s6-finish] sending all processes the TERM signal.
srun: error: sh-06-34: task 0: Hangup

$ hostname
sh-ln01.int
```

The processes within the container are killed, the singularity process is killed, the SSH/interactive job session is killed and I'm disconnected.

Try this as root, and the `sshd` daemon is killed.

Try this in a desktop  environment, and the X11 session is killed.

This sounds like a major problem to me.

**Comment by vsoch** [comment]
To add some, I have tried simple running of the above container, and then building (with the [/init sequence](https://github.com/rocker-org/rocker-versioned/blob/master/rstudio/Dockerfile#L92)) as sudo, and both times my computer becomes unresponsive, crashes, and once I was able to get a desktop, but all network connectivity was gone.  I don't dare debug this again because it crashes *every single time*. If you dare give it a look, here is the Singularity file I was using (to debug the docker above):

```
Bootstrap: docker
From: rocker/rstudio

# sudo singularity build r-singularity

%files
    . /home/rstudio/project

%post

    # install packrat

    R -e 'install.packages("packrat", repos="http://cran.rstudio.com", dependencies=TRUE, lib="/usr/local/lib/R/site-library");'

    # copy lock file & install deps
    R -e 'packrat::restore(project="/home/rstudio/project");'
   
    chown -R rstudio /home/rstudio

    # Init the files needed in the container
    /bin/bash /init
```
Actually I can do better, here is the entire repository, the "singularity" branch --> https://github.com/vsoch/r-docker-ex/tree/singularity  

You'll notice that the tutorial cuts short... there is a reason for that, above :) May the force be with you!

**Comment by dtrudg** [comment]
TLDR: Run using the `-p` option to `singularity run`


Hi @kcgthb @vsoch - this is not necessarily a bug, but an interaction between this container, and Singularity's default level of isolation.

In the rstudio container, the s6 init / process supervisor system is being used, from the `s6-overlay` distribution. This is configured so that when the supervised (rstudio) process exits, `s6-nuke` will be called - which tries to kill every process it can see. 

By default, Singularity containers do not run in a separate process namespace - if you do a `ps` inside a container you will see processes outside. This is one of the ways Singularity differs from docker, and why the `s6-nuke` that is invoked is killing the processes as you see.

You can ask Singularity to use a separate pid namespace by running it with the `-p` option. If you do a `ps` inside a container run with `-p` you will only see processes inside the container. If you run the rstudio container with `-p` it should not kill anything on the host via `s6-nuke` when you end it.

**Comment by vsoch** [comment]
awesome! I'm trying this out now (with shell first).

**Comment by vsoch** [comment]
okay, for the shell it didn't whack out, I'm guessing the same would happen for run/exec and I'm going to write a more thorough recipe for a container and I'll post it here.

**Comment by kcgthb** [comment]
Hi @dctrud,

Can `-p` be enabled by default? Maybe with an option in the `singularity.conf` file?

What are the side effects of running containers in a separate process space, with things like MPI applications, for instance?

I'm thinking that this behavior is sufficitently dangerous and disturbing that process isolation should be enabled by default, with a runtime flag to disable it, rather than the other way around as it is now.

What do you think?

**Comment by dtrudg** [comment]
Hi @kcgthb - the explanation for the behaviour is here: http://singularity.lbl.gov/docs-config

>While the PID namespace is a neat feature, it does not have much practical usage in an HPC context so it is recommended to disable this if you are running on an HPC system where a resource manager is involved as it has been known to cause confusion on some kernels with enforcement of user limits.

>Even if the PID namespace is enabled by the system administrator here, it is not implemented by default when running containers. The user will have to specify they wish to implement un-sharing of the PID namespace as it must fork a child process.

**Comment by dtrudg** [comment]
Hi all - as this has been open for some time without follow up I'm going to close. Feel free to reopen if there's anything else.

**Comment by vsoch** [comment]
If this is rstudio, then this is some flavor of the error reported [here](https://github.com/sylabs/singularity/issues/3777). The resolution (for rstudio) is not to use their `/init`, but just start rstudio server with `rstudio-server start`. If anyone else ever comes across this, never run rstudio (from Docker) out of the box, and also @kcgthb purpose (getting Rstudio working with Singularity) is now one of the provided apps for Ohio Supercomputer On Demand [see here](https://github.com/OSC/bc_osc_rstudio_server/blob/master/template/script.sh.erb#L58) so try that first, if you need Rstudio Server on an HPC cluster.
